saas/models/saas_template.py

```diff
@@ -0,0 +1,143 @@
+        'Modules to install',
+        help='Domain to search for modules to install after template database creation',
+        defalt="[]")
+    template_post_init = fields.Text(
+        'Template Initialization',
+        default=lambda s: s.env['ir.actions.server'].DEFAULT_PYTHON_CODE,
+        help='Python code to be executed once db is created and modules are installed')
+    operator_ids = fields.One2many(
+        'saas.template.operator',
+        'template_id')
+
+    @api.constrains('template_post_init')
+    def _check_python_code(self):
+        for r in self.sudo():
+            msg = test_python_expr(expr=r.template_post_init.strip(), mode="exec")
+            if msg:
+                raise ValidationError(msg)
+
+
+class SAASTemplateLine(models.Model):
+    _name = 'saas.template.operator'
+    _description = 'Template\'s Settings for Operator'
+
+    template_id = fields.Many2one('saas.template', required=True)
+    operator_id = fields.Many2one('saas.operator', required=True)
+    password = fields.Char('DB Password')
+    operator_db_name = fields.Char(required=True)
+    operator_db_id = fields.Many2one('saas.db', readonly=True)
+    operator_db_state = fields.Selection(related='operator_db_id.state')
+    to_rebuild = fields.Boolean()
+    state = fields.Selection([
+        ('draft', 'Draft'),
+        ('creating', 'Database Creating'),
+        ('installing_modules', 'Modules installation'),
+        ('post_init', 'Extra initialization'),
+        ('done', 'Ready'),
+
+    ])
+
+    def preparing_template_next(self):
+        # TODO: This method is called by cron every few minutes
+        template_operators = self.search([('to_rebuild', '=', True)])
+        operators = template_operators.mapped('operator_id')
+
+        # filter out operators which already have template creating
+        def filter_free_operators(op):
+            states = op.template_operator_ids.mapped('state')
+            return all((s in ['draft', 'done'] for s in states))
+
+        operators = operators.filtered(filter_free_operators)
+        if not operators:
+            # it's not a time to start
+            return
+
+        for t_op in template_operators:
+            if t_op.operator_id not in operators:
+                continue
+            t_op._prepare_template()
+
+            # only one template per operator
+            operators -= t_op.operator_id
+
+    def _prepare_template(self):
+        for r in self:
+            # delete db is there is one
+            r.operator_db_id.drop_db()
+            if not r.operator_db_id or r.operator_id != r.operator_db_id.operator_id:
+                r.operator_db_id = self.env['saas.db'].create({
+                    'name': r.operator_db_name,
+                    'operator_id': r.operator_id.id,
+                    'type': 'template',
+                })
+            password = random_password()
+            self.env['saas.log'].log_db_creating(self)
```

Comment:
It is strange that this piece of code just now started to generate an error. The log_db_creating takes instance of saas.db

Reply:
Can you fix the arguments or the method?

Reply:
I already fixed it. I'll post changes soon